### ISSUE TITLE
Add clear color to Viewports

### ISSUE THREAD

**Issue by BenMcLean** [godotengine/godot-proposals#252]
**Describe the project you are working on:**

Trying to put a virtual computer screen with old school computer text on it inside a 3D game.

**Describe the problem or limitation you are having in your project:**

It isn't readily apparent where the Viewport gets its clear color from but evidently there isn't a way to set a different clear color on different Viewports.

Right now, the workaround for not having a way to set a clear color for individual Viewports is adding a ColorRect to its tree. I can use that for now but it is still annoying.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

There should be a property on the Viewport to let you set a default clear color. That would make Viewports easier to work with instead of requiring another object to clear the viewport with.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

A whole additional seemingly unnecessary object is required to do something Viewports should do on their own.

**Is there a reason why this should be core and not an add-on in the asset library?:**

Lacking this feature seems like an oversight.

**Comment by clayjohn** [comment]
Background color is set within the environment of a Viewport. Near the top of the inspector there is a property for a Viewport to use its own world. Select that, assign an environment, and then set the background color there.

**Comment by golddotasksquestions** [comment]
Sounds to me like OP is talking also (or exclusively) about the 2D viewport @clayjohn .
I first thought that's probably a color you could set in Editor Settings>Editor>2D, but I could not find it. Right clicking also does nothing.
I then checked in the World environment node, but "Canvas" mode does not seem to have a setting for the viewport background color and setting the mode to "CustomColor" changes nothing in 2D.

**Comment by BenMcLean** [comment]
Yes that is what I was trying to say.

**Comment by TheWhiteLlama** [comment]
Yea, as said by clayjohn the clear color is defined by an environment. While there's always a default environment which is defined in the project's settings, each viewport will either use the default environment or the one defined in the node inspector.

So what I basically want to point out here:
The clear color is **always** defined by an environment at the moment.

Adding another property in the inspector might even confuse beginners since it wouldn't be obvious if the clear color was defined by the property or the environment for the viewport

**Comment by girng** [comment]
`set_default_clear_color`?

I don't see an issue adding it to the VP's inspector.. that would be a good idea. It is kind of hidden

**Comment by clayjohn** [comment]
@girng it would have to be "override_background_color" because background color is set based on environment, not Viewport.

**Comment by girng** [comment]
TIL. I always thought it was just the VP's bg color, ty for the clarification. Okay that's probably why it's not in the VP's inspector then

**Comment by seocwen** [comment]
Environment doesn't work for 2D games. 2D games often require their own background color, and having to stretch a color rect is hacky. You can set the background color for ALL viewports in the project preferences, (or by calling VisualServer.set_default_clear_color()) but it's a very common use case that you would want to set the background color for EACH viewport you are using.

**Comment by clayjohn** [comment]
We discussed this at today's proposal review meeting. We agreed on an implementation as follows:
1. Add a per-viewport setting for clear color (could be renamed background color)
2. Replace the current implementation of clear color to use Viewport clear color instead (Project setting would remain, but would set the default clear color for each viewport instead, each viewport could then change their clear color in editor or at run time)
3. Internally this would work the same from the renderer's perspective, if no Environment is present, or if background mode is set to clear_color, the Viewport's clear color will be used.

The implementation of the above should not be too complex. The difficult part of this proposal will be ensuring that nothing breaks in the editor as it makes heavy use of Viewports

**Comment by golddotasksquestions** [comment]
Yes please rename this to "default_background_color" (for the project settings) and "custom_background_color" instead of "clear_color". A clear_color to me is a color with 100% transparency. This makes no sense to me here since I can clearly see the color.

**Comment by clayjohn** [comment]
> Yes please rename this to `default_background_color` (for the project settings) and `custom_background_color` instead of `clear_color`. A clear_color to me is a color with 100% transparency. This makes no sense to me here since I can clearly see the color.

That makes sense!

**Comment by YuriSizov** [comment]
The rename makes sense for, uhm, clarity, but just for the sake of, eh, transparency — it is called a "clear" color on the graphic library level, as it is the color used to clear the rendering surface. I.e. it is the color that you wipe the screen with.

**Comment by Calinou** [comment]
If we add a per-viewport clear color option, there are two questions:

- Changing the Clear Color on a per-viewport basis should affect radiance map generation if possible.
- Should Environment keep its Custom Color property? Remember that it also affects the radiance map generation.

**Comment by golddotasksquestions** [comment]
> The rename makes sense for, uhm, clarity, but just for the sake of, eh, transparency — it is called a "clear" color on the graphic library level, as it is the color used to clear the rendering surface. I.e. it is the color that you wipe the screen with.

That's super interesting fact, thanks you for sharing! Unfortunately naming like this is usually completely lost to the end user who works and experiences the tool on a much higher level.

**Comment by clayjohn** [comment]
> * Changing the Clear Color (background color) on a per-viewport basis should affect radiance map generation if possible. Is it possible to do this on a per-viewport basis, though?

Well, the background color doesn't impact radiance map generation. We already store a clear_color per rendertarget internally. The only change is that we would remove the default clear color and instead always rely on clear color set per Viewport/RenderTarget. Everything else with respect to 3D rendering/the environment can stay the same.

> * Should Environment keep its Custom Color property? Remember that it also affects the radiance map generation.

Environment doesn't need to change. This proposal will only affect the environment when background mode clear color is used. In that case, the per-viewport clear color will be used.

**Comment by jeffbdavenport** [comment]
> Background color is set within the environment of a Viewport. Near the top of the inspector there is a property for a Viewport to use its own world. Select that, assign an environment, and then set the background color there.

Yeah this does nothing in 4.2...

**Comment by balloonpopper** [comment]
I've been struggling to work out which setting controls the background color of a window node for the last hour (adding pretty much every node type to the theme and changing its colours). It's not intuitive that everything else about a window is available in the theme, but its background is covered in Godot's viewport settings. A theme setting would be intuitive.

If there's a property somewhere (as mentioned above) under the "window" node type that's related to "viewport" or "world" to set its colour under 4.3 beta 2, I can't find it.

**Comment by Hydra9268** [comment]
It seems like there's definitely a need to set the environment's background color at the scene level, and it would make a lot of sense if this could be done using a color value—especially if it were something like `@export var`. That way, it'd be super easy to tweak per scene without digging into extra menus or overrides.

`@export var bg_color: Color = Color("ff0000")`

So now you can set a scene's background color independently from other scenes, either directly in code or through the Inspector.
